oauth2_http/javatests/com/google/auth/oauth2/JwtCredentialsTest.java

```diff
@@ -0,0 +1,302 @@
+      ServiceAccountJwtAccessCredentials.JWT_ACCESS_PREFIX;
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  static PrivateKey getPrivateKey() {
+    try {
+      return ServiceAccountCredentials.privateKeyFromPkcs8(PRIVATE_KEY);
+    } catch (IOException ex) {
+      return null;
+    }
+  }
+
+  @Test
+  public void serialize() throws IOException, ClassNotFoundException {
+    JwtCredentials.Claims claims = JwtCredentials.Claims.newBuilder()
+        .setAudience("some-audience")
+        .setIssuer("some-issuer")
+        .setSubject("some-subject")
+        .build();
+    JwtCredentials credentials = JwtCredentials.newBuilder()
+        .setClaims(claims)
+        .setPrivateKey(getPrivateKey())
+        .setPrivateKeyId(PRIVATE_KEY_ID)
+        .build();
+
+    JwtCredentials deserializedCredentials = serializeAndDeserialize(credentials);
+    assertEquals(credentials, deserializedCredentials);
+    assertEquals(credentials.hashCode(), deserializedCredentials.hashCode());
+    assertEquals(credentials.toString(), deserializedCredentials.toString());
+    assertSame(deserializedCredentials.getClock(), Clock.SYSTEM);
+  }
+
+  @Test
+  public void builder_requiresPrivateKey() {
+    try {
+      JwtCredentials.Claims claims = JwtCredentials.Claims.newBuilder()
+          .setAudience("some-audience")
+          .setIssuer("some-issuer")
+          .setSubject("some-subject")
+          .build();
+      JwtCredentials.newBuilder()
+          .setClaims(claims)
+          .setPrivateKeyId(PRIVATE_KEY_ID)
+          .build();
+      fail("Should throw exception");
+    } catch (NullPointerException ex) {
+      // expected
+    }
+  }
+
+  @Test
+  public void builder_requiresPrivateKeyId() {
+    try {
+      JwtCredentials.Claims claims = JwtCredentials.Claims.newBuilder()
+          .setAudience("some-audience")
+          .setIssuer("some-issuer")
+          .setSubject("some-subject")
+          .build();
+      JwtCredentials.newBuilder()
+          .setClaims(claims)
+          .setPrivateKey(getPrivateKey())
+          .build();
+      fail("Should throw exception");
+    } catch (NullPointerException ex) {
+      // expected
+    }
+  }
+
+  @Test
+  public void builder_requiresClaims() {
+    try {
+      JwtCredentials.newBuilder()
+          .setPrivateKeyId(PRIVATE_KEY_ID)
+          .setPrivateKey(getPrivateKey())
+          .build();
+      fail("Should throw exception");
+    } catch (NullPointerException ex) {
+      // expected
+    }
+  }
+
+  @Test
+  public void builder_requiresCompleteClaims() {
+    try {
+      JwtCredentials.Claims claims = JwtCredentials.Claims.newBuilder()
+          .build();
+      JwtCredentials.newBuilder()
+          .setClaims(claims)
+          .setPrivateKeyId(PRIVATE_KEY_ID)
+          .setPrivateKey(getPrivateKey())
+          .build();
+      fail("Should throw exception");
+    } catch (IllegalStateException ex) {
+      // expected
```

Comment:
why IllegalState here and NullPointer above? They all feel like IllegalState to me.

Reply:
Implementation-wise, it's using `Preconditions.checkState()` to ensure that all required claim fields are correctly set. It's not a NullPointerException because the claims are provided, they are just incomplete.